src/wwwroot/app/views/reports/users-report/users.report.controller.js

```diff
@@ -0,0 +1,225 @@
+            }
+         });
+      });
+   }());
+
+   function useLocationField(isActiveLocationField) {
+      if (!isActiveLocationField) {
+         _clearLocationField();
+      }
+   }
+
+   function useUserField(isActiveUsersField) {
+      if (!isActiveUsersField) {
+         _clearUserField();
+      }
+   }
+
+   function _clearLocationField() {
+      vm.usersReportParametrs.locationIds = [];
+      vm.selectedLocations = [];
+      vm.selectedUsersGroupedByLocation = {};
+      each(vm.locations, (location) => {
+         location.selected = false;
+      });
+   }
+
+   function _clearUserField() {
+      vm.usersReportParametrs.userIds = [];
+      vm.selectedUsers = [];
+      vm.selectedUsersGroupedByLocation = {};
+      each(vm.users, (user) => {
+         user.selected = false;
+      });
+   }
+
+   function addLocationIdsToUsersReportParametrs(location) {
+      if (location.selected && !vm.usersReportParametrs.locationIds.includes(location.id)) {
+         vm.usersReportParametrs.locationIds.push(location.id);
+         vm.selectedLocations.push(location);
+      } else if (!location.selected) {
+         remove(vm.usersReportParametrs.locationIds, (locationId) =>  locationId === location.id);
+         remove(vm.selectedLocations, (loc) =>  loc.id === location.id);
+      }
+   }
+
+   function addUserIdsToUsersReportParametrs(user) {
+      if (user.selected && !vm.usersReportParametrs.userIds.includes(user.id)) {
+         vm.usersReportParametrs.userIds.push(user.id);
+         if (vm.selectedLocations.length) {
+            _convertUsersArrayToHash(user);
+         } else {
+            vm.selectedUsers.push(user);
+         }
+      } else if (!user.selected) {
+         remove(vm.usersReportParametrs.userIds, (userId) =>  userId === user.id);
+         if (vm.selectedLocations.length) {
+            remove(vm.selectedUsersGroupedByLocation[user.cityId], {id: user.id});
+         } else {
+            remove(vm.selectedUsers, (us) =>  us.id === user.id);
+         }
+      }
+   }
+
+   function formingUsersReport(form) {
+      if (!vm.selectedLocations.length && !vm.selectedUsers.length) {
+         UserDialogService.notification($translate.instant('DIALOG_SERVICE.EMPTY_REPORT_CONDITIONS'), 'error');
+         return false;
+      }
+      if (vm.usersReportParametrs.startDate > vm.usersReportParametrs.endDate) {
+         UserDialogService.notification($translate.instant('DIALOG_SERVICE.INVALID_DATES'), 'error');
+         return false;
+      }
+      ValidationService.validate(form).then(() => {
+         ReportsService.getDataForUserReport(vm.usersReportParametrs).then(resp => {
+            vm.usersReportParametrs.startDate = resp.startDate;
+            vm.usersReportParametrs.endDate = resp.endDate;
+            _convertReport(resp.userReport);
+         });
+      });
+      return false;
+   }
+
+   function sumReportByStages(report, stageId) {
+      if (report !== undefined) {
```

Comment:
If report will be indefined, we cannot read property stagesData[stageId] of undefined